### ISSUE TITLE
Improve safety of assigned ordinals

### ISSUE THREAD

**Issue by DanielThomas** [Netflix/hollow#301]
It's possible to accidentally misuse the `__assigned_ordinal` field:

- If it's defined as `int`, the value is used as the ordinal
- If the value is accidentally final, it fails silently (very easy to do with Kotlin `val`)
- `HollowConstants` defines `ORDINAL_NONE` as an `int` which could lead to confusion (and makes Kotlin unhappy - it wants you to be explicit about int -> long conversions)
```
    /**
     * An ordinal of NULL_ORDINAL signifies "null reference" or "no ordinal"
     */
    int ORDINAL_NONE = -1;
```

**Comment by DanielThomas** [comment]
Added some tests to demonstrate the issue - https://github.com/Netflix/hollow/compare/issue-301.

**Comment by dkoszewnik** [comment]
@PaulSandoz In Hollow v4, please consider removing support for `int` `__assigned_ordinal` fields.  These should ideally always be `long`.

**Comment by PaulSandoz** [comment]
@dkoszewnik +1 I agree this functionality should be removed

**Comment by PaulSandoz** [comment]
#317 was merged into the hollow4 branch

**Comment by PaulSandoz** [comment]
Fixed in #317
